File: src/test/java/io/mfedirko/common/infra/dynamodb/DynamoDbLearningRepositoryTest.java
package io.mfedirko.common.infra.dynamodb;

import io.mfedirko.DynamoDbTestConfiguration;
import io.mfedirko.learning.CreateLessonForm;
import io.mfedirko.learning.Lesson;
import io.mfedirko.learning.UpdateLessonForm;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;

import java.time.Year;
import java.util.List;

import static io.mfedirko.RepositoryTestHelpers.sortedDescending;
import static io.mfedirko.RepositoryTestHelpers.withinYear;

@SpringBootTest
@Import({DynamoDbTestConfiguration.class})
class DynamoDbLearningRepositoryTest {
    @Autowired
    private DynamoDbLearningRepository repository;


    @Nested
    class SaveLessons {
        @Test
        void updateExisting() {
            Year year = Year.of(2023);
            Lesson lesson = repository.findLessons(year)
                    .get(0);
            UpdateLessonForm updateForm = UpdateLessonForm.builder()
                    .category(lesson.getCategory())
                    .description("MY NEW DESCRIPTION OF LESSON!")
                    .title("MY NEW TITLE of lessoin")
                    .build();

            repository.updateLesson(updateForm, lesson.getCreationTimestampMillis());

            List<Lesson> lessons = repository.findLessons(year);
            Assertions.assertThat(lessons)
                    .filteredOn(l -> l.getCreationTimestamp().equals(lesson.getCreationTimestamp()))
                    .hasSize(1)
                    .allMatch(l -> l.getDescription().equals(updateForm.getDescription())
                            && l.getTitle().equals(updateForm.getTitle()));
        }

        @Test
        void updateMergesOldFields() {
            Year year = Year.of(2023);
            Lesson lesson = repository.findLessons(year)
                    .get(0);
            Assertions.assertThat(lesson.getAuthor()).isNotEmpty();

            UpdateLessonForm updateForm = UpdateLessonForm.builder()
                    .title("MY NEW TITLE of lessoin")
                    .build();

            repository.updateLesson(updateForm, lesson.getCreationTimestampMillis());

            List<Lesson> lessons = repository.findLessons(year);
            Assertions.assertThat(lessons)
                    .filteredOn(l -> l.getCreationTimestamp().equals(lesson.getCreationTimestamp()))
                    .hasSize(1)
                    .allSatisfy(l -> Assertions.assertThat(l).hasNoNullFieldsOrProperties()
                            .hasFieldOrPropertyWithValue("title", updateForm.getTitle()));
        }


        @Test
        void deleteLesson() {
            Year year = Year.of(2023);
            Lesson lesson = repository.findLessons(year)
                    .get(0);

            repository.deleteLesson(lesson.getCreationTimestampMillis());

            List<Lesson> lessons = repository.findLessons(year);
            Assertions.assertThat(lessons)
                    .filteredOn(l -> l.getCreationTimestamp().equals(lesson.getCreationTimestamp()))
                    .isEmpty();
        }

        @Test
        void createNew() {
            CreateLessonForm form = CreateLessonForm.builder()
                    .category("NEW CATEGORY")
                    .description("MY EXPETED DESCRIPTION")
                    .title("MY TITLE FOR NEW LESSON")
                    .build();

            repository.createLesson(form);

            List<Lesson> lessons = repository.findLessons(Year.now());
            Assertions.assertThat(lessons)
                    .filteredOn(l -> l.getCategory().equals(form.getCategory())
                        && l.getDescription().equals(form.getDescription())
                        && l.getTitle().equals(form.getTitle()))
                    .hasSize(1);
        }
    }

    @Nested
    class FindLessons {
        @ParameterizedTest
        @ValueSource(strings = {
                "2023",
                "2022"
        })
        void withinRange(Year year) {
            List<Lesson> lessons = repository.findLessons(year);

            Assertions.assertThat(lessons).isNotEmpty();
        }

        @ParameterizedTest
        @ValueSource(strings = {
                "2023",
                "2022"
        })
        void withinRangeOfYear(Year year) {
            List<Lesson> lessons = repository.findLessons(year);

            Assertions.assertThat(lessons).isNotEmpty().are(withinYear(year.getValue(), Lesson::getCreationTimestamp));
        }

        @ParameterizedTest
        @ValueSource(strings = {
                "2023",
                "2022"
        })
        void sortedDescendingByTimestamp(Year year) {
            List<Lesson> lessons = repository.findLessons(year);

            Assertions.assertThat(lessons).is(sortedDescending(Lesson::getCreationTimestamp));
        }
    }


}

File: src/test/kotlin/io/mfedirko/common/infra/dynamodb/DynamoDbLearningRepositoryTest.kt
package io.mfedirko.common.infra.dynamodb

import io.mfedirko.RepositoryTestConditions.sortedDescending
import io.mfedirko.RepositoryTestConditions.withinYear
import io.mfedirko.learning.CreateLessonForm
import io.mfedirko.learning.Lesson
import io.mfedirko.learning.UpdateLessonForm
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import java.time.Year

@SpringBootTest(classes = [
    DynamoDbTestConfiguration::class,
    DynamoDbLearningRepository::class
])
@ActiveProfiles("aws")
internal class DynamoDbLearningRepositoryTest {
    @Autowired
    private lateinit var repository: DynamoDbLearningRepository

    @Nested
    internal inner class SaveLessons {
        @Test
        fun updateExisting() {
            val year = Year.of(2023)
            val lesson = repository.findLessons(year)[0]
            val updateForm = UpdateLessonForm().apply {
                category = lesson.category
                description = "MY NEW DESCRIPTION OF LESSON!"
                title = "MY NEW TITLE of lessoin"
            }
            repository.updateLesson(updateForm, lesson.id)
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons)
                .filteredOn { it.creationTimestamp == lesson.creationTimestamp }
                .hasSize(1)
                .allMatch { it.description == updateForm.description && it.title == updateForm.title }
        }

        @Test
        fun updateMergesOldFields() {
            val year = Year.of(2023)
            val lesson = repository.findLessons(year)[0]
            Assertions.assertThat(lesson.author).isNotEmpty
            val updateForm = UpdateLessonForm.fromLesson(lesson).apply {
                title = "MY NEW TITLE of lessoin"
            }
            repository.updateLesson(updateForm, lesson.id)
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons)
                .filteredOn { it.creationTimestamp == lesson.creationTimestamp }
                .hasSize(1)
                .allSatisfy { lsn: Lesson ->
                    Assertions.assertThat(lsn).hasNoNullFieldsOrProperties()
                        .hasFieldOrPropertyWithValue("title", updateForm.title)
                }
        }

        @Test
        fun deleteLesson() {
            val year = Year.of(2023)
            val lesson = repository.findLessons(year)[0]
            repository.deleteLesson(lesson.id)
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons)
                .filteredOn { it.creationTimestamp == lesson.creationTimestamp }
                .isEmpty()
        }

        @Test
        fun createNew() {
            val form = CreateLessonForm().apply {
                category = "NEW CATEGORY"
                description = "MY EXPETED DESCRIPTION"
                title = "MY TITLE FOR NEW LESSON"
            }
            repository.createLesson(form)
            val lessons = repository.findLessons(Year.now())
            Assertions.assertThat(lessons)
                .filteredOn { l: Lesson -> l.category == form.category && l.description == form.description && l.title == form.title }
                .hasSize(1)
        }
    }

    @Nested
    internal inner class FindLessons {
        @ParameterizedTest
        @ValueSource(strings = ["2023", "2022"])
        fun withinRange(year: Year) {
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons).isNotEmpty
        }

        @ParameterizedTest
        @ValueSource(strings = ["2023", "2022"])
        fun withinRangeOfYear(year: Year) {
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons).isNotEmpty.are(withinYear(year.value) { it.creationTimestamp })
        }

        @ParameterizedTest
        @ValueSource(strings = ["2023", "2022"])
        fun sortedDescendingByTimestamp(year: Year) {
            val lessons = repository.findLessons(year)
            Assertions.assertThat(lessons)
                .`is`(sortedDescending { it.creationTimestamp })
        }
    }
}